### ISSUE TITLE
MySQL SSH error: com.mysql.jdbc.exceptions.jdbc4.CommunicationsException: Communications link failure

### ISSUE THREAD

**Issue by vishal-reddy** [metabase/metabase#3941]
### Bug

- Chrome 54
- macOS Sierra
- MySQL 5.7.16 Homebrew
- Metabase version: v0.20.3
- Metabase hosting environment: (Docker)
- Metabase internal database: ( (default))

#### Reproduce

Attempting to Add database 

Database type: MySQL

Name: 
LocalLiferay
Host: 
localhost
Port: 
3306
Database name: 
lportal
Database username: 
root
Database password: 
••••

Enable in-depth database analysis
This allows us to present you with better metadata for your tables and is required for some features of Metabase. We recommend leaving this on unless your database is large and you're concerned about performance.
Save

Getting this error below: 

No matching clause: com.mysql.jdbc.exceptions.jdbc4.CommunicationsException: Communications link failure The last packet sent successfully to the server was 0 milliseconds ago. The driver has not received any packets from the server.

I am able to access DB from command line.

:arrow_down: Please click the :+1: reaction instead of leaving a `+1` or `update?` comment

**Comment by camsaul** [comment]
Hi @vishal-reddy, that sounds like it could be an issue with the MySQL JDBC driver or somethings else rather than Metabase itself. Are you able to connect to your DB with other Java-based tools? Can Metabase connect to any of your other MySQL DBs?

**Comment by vishal-reddy** [comment]
Yes. I am able to connect with the Intellij Database tool which uses JDBC

**Comment by jmmk** [comment]
I have the same issue with a local brew-installed MySQL with both JDBC and .NET MySQL drivers, but C and Python drivers work fine.

I was able to get it working by dumping my data and loading it into MySQL running on docker (running Ubuntu). Nothing I tried worked when trying to connect to the Homebrew MySQL. I think either Sierra or Homebrew MySQL for Sierra must have changed a setting somewhere that causes the drivers to malfunction because I have used JDBC heavily in the past with no issues.

**Comment by vishal-reddy** [comment]
Thanks @jmmk. Tried to connect to a remote MySQL running on a Linux server. Same error but even that MySQL server is able to connect from Intellij as well.

**Comment by yanghaogn** [comment]
same problem

> No matching clause: com.mysql.jdbc.exceptions.jdbc4.CommunicationsException: Communications link failure The last packet sent successfully to the server was 0 milliseconds ago. The driver has not received any packets from the server.

**Comment by Poggen** [comment]
We're also experiencing this, i.e. users regularly experience error messages such as
>"The last packet successfully received from the server was 845 828 milliseconds ago.  The last packet sent successfully to the server was 0 milliseconds ago."

Our theory is that the connection to the database is kept open for too long. Our MySQL server has  `wait_timeout=600` `interactive_timeout=600` and `lock_wait_timeout=120`. Since 845 sec > 600 sec, the sql server has already closed the connection, resulting in the error above. Any idea why metabase isn't reconnecting to the database properly?

**Comment by cdmarioe** [comment]
Same Error

**Comment by jonnyenglish** [comment]
Same Error.
I have two docker containers running. One for mysql, one for metabase. Everything runs fine, just every hour in the logs I get this error. (No users on this instance)

**Comment by cdmarioe** [comment]
I had to use Java version..

**Comment by jonnyenglish** [comment]
@cdmarioe what do you mean? Have you been able to resolve this issue?

**Comment by cdmarioe** [comment]
My first try was with Docker, i wasn't able to run metabase...the second one, i downloaded metabase.jar and runs perfect..

**Comment by slehmann36** [comment]
I have the same issue... Are there any updates or potential fixes for this?

**Comment by joebordes** [comment]
I just ran into this while installing docker container of metabase. I had to open mysql to listen on all networks (/etc/mysql/my.cnf: bind=0.0.0.0) and then grant access to the database user from the IP assigned to the docker container.
That solved the issue and I can access the database.

HTH

**Comment by mivbeard** [comment]
I have the same error

**Comment by incyclum** [comment]
@joebordes I tried but It didn't worked for me. Changed bind-adress to 0.0.0.0, created a mysql user user@container-ip but it doesn't work. have the same error message.

**Comment by joebordes** [comment]
Have you restarted MySQL service?

My docker network is 172.17.0.*, I am accessing as root user and this is what my permissions look like:

![mysqlgrant](https://user-images.githubusercontent.com/1237191/27042409-2166d5c8-4f97-11e7-949d-de006805a549.png)

**Comment by incyclum** [comment]
I think you have your docker container on the same host than mysql right (which is not containerized) ?

I have a metabase container on host1 and mysql (non containerized) on host2. I want to connect over ssh.

**Comment by joebordes** [comment]
Correct, MySQL is on host, not in docker container.

Can you connect via SSH tunnel from the command line in the metabase container?

**Comment by incyclum** [comment]
The metabase container does not contain the ssh binary :/

**Comment by joebordes** [comment]
I installed the ssh client inside the container with these two commands:

```
apk update
apk add openssh
```

and was able to ssh out to my host machine

**Comment by jonnylangefeld** [comment]
Since I still got e-mails of this thread I wanted to post my solution here:

My setup is multiple docker containers one of which the official metabase one and one of which a mysql one. 
I found out that the error message for me occurred every hour. That is because metabase is configured with batch jobs being executed every hour (I think one is for e-mails and I forgot about the other one)
For these batch jobs metabase was requesting the mysql with the connection it opened when I started metabase (so from one hour ago). It did not start a new connection. The default connection time out of mysql is 600 sec. 
So after 10 mins mysql closed the existing connection because it was not used, but 50 minutes later metabase tried to use this exact same connection.
My solution was just to increase the default connection wait_timeout in `.etc/mysql/conf.d/custom.cnf` to

    [mysqld]
    wait_timeout		= 11000

Hope this helps someone.

**Comment by incyclum** [comment]
I suppose I'm already sshing successfuly between the hosts, because If i change the host or credentials I have an SSH related error message.

**Comment by stoneo** [comment]
My temporary solution : 

```
apk add --update openssh
ssh -fN -L 3306:localhost:3306 user@server -p 22
```

**Comment by lazmo88** [comment]
@ston3o This worked for me as well! Thanks :) I was able to run tunnel connection from my local computer to the remote server in my initial testing, however when I installed Metabase to my cloud server and tried using the GUI for setting SSH connection, it failed as mentioned by OP. However manually creating ssh tunnel to the remote server worked. I believe there is something wrong with Metabase GUI / SSH...

**Comment by jun85664396** [comment]
@Poggen 
I solved the problem.
add `interactiveClient=true` in the JDBC connection options.
![image](https://user-images.githubusercontent.com/7875341/35089399-2e86d420-fc7a-11e7-9d64-b10d4a32eae6.png)

**Comment by Silvertale** [comment]
For anyone coming here because you are running in docker: You need to set the bind-address to 0.0.0.0 in your mysql configs on your host, restart mysql, and add a mysql user with 'MY_USERNAME'@'THE_DOCKER_CONTAINER_IP_ADDRESS'

**Comment by d38u993** [comment]
> 
> 
> @Poggen
> I solved the problem.
> add `interactiveClient=true&autoReconnectForPools=true` in the JDBC connection options.
> or
> execute `show variables` in the metabase question and check the timeout variables.

Gracias si me funciono, parece que solo me faltaba esto. ya que esto lo hice tambien bind-address to 0.0.0.0
y me hice un nuevo usuario en la base de datos de MySQL pero le quite la contraseña, tambien comente una opcion que dice skip-networking....

**Comment by jeff303** [comment]
> For anyone coming here because you are running in docker: You need to set the bind-address to 0.0.0.0 in your mysql configs on your host, restart mysql, and add a mysql user with 'MY_USERNAME'@'THE_DOCKER_CONTAINER_IP_ADDRESS'

Yes, MySQL is picky about both the username and "host" the client is connecting from.

**Comment by flamber** [comment]
Closing, since SSH tunnel reconnect was fixed in 0.39.0 #10081 and connection timeout has been extended in 0.41.1 #18354

If you are still experiencing this problem, then post "Diagnostic Info" from Admin > Troubleshooting, and the full stacktrace from Admin > Troubleshooting > Logs. And any other details that would be helpful to reproduce the problem.
